text/0000-api-evolution.md

```diff
@@ -0,0 +1,798 @@
+
+[Rust 1.0.0](http://blog.rust-lang.org/2015/02/13/Final-1.0-timeline.html) will
+mark the beginning of our
+[commitment to stability](http://blog.rust-lang.org/2014/10/30/Stability.html),
+and from that point onward it will be important to be clear about what
+constitutes a breaking change, in order for semver to play a meaningful role. As
+we will see, this question is more subtle than one might think at first -- and
+the simplest approach would make it effectively impossible to grow the standard
+library.
+
+The goal of this RFC is to lay out a comprehensive policy for what *must* be
+considered a breaking API change from the perspective of semver, along with some
+guidance about non-semver-breaking changes.
+
+# Detailed design
+
+For clarity, in the rest of the RFC, we will use the following terms:
+
+* **Major change**: a change that requires a major semver bump.
+* **Minor change**: a change that requires only a minor semver bump.
+* **Breaking change**: a change that, *strictly speaking*, can cause downstream
+  code to fail to compile.
+
+What we will see is that in Rust today, almost any change is technically a
+breaking change. For example, given the way that globs currently work, *adding
+any public item* to a library can break its clients (more on that later). But
+not all breaking changes are equal.
+
+So, this RFC proposes that **all major changes are breaking, but not all breaking
+changes are major.**
+
+## Overview
+
+### Principles of the policy
+
+The basic design of the policy is that **the same code should be able to run
+against different minor revisions**. Furthermore, minor changes should require
+at most a few local *annotations* to the code you are developing, and in
+principle no changes to your dependencies.
+
+In more detail:
+
+* Minor changes should require at most minor amounts of work upon upgrade. For
+  example, changes that may require occasional type annotations or use of UFCS
+  to disambiguate are not automatically "major" changes. (But in such cases, one
+  must evaluate how widespread these "minor" changes are).
+
+* In principle, it should be possible to produce a version of dependency code
+  that *will not break* when upgrading other dependencies, or Rust itself, to a
+  new minor revision. This goes hand-in-hand with the above bullet; as we will
+  see, it's possible to save a fully "elaborated" version of upstream code that
+  does not require any disambiguation. The "in principle" refers to the fact
+  that getting there may require some additional tooling or language support,
+  which this RFC outlines.
+
+That means that any breakage in a minor release must be very "shallow": it must
+always be possible to locally fix the problem through some kind of
+disambiguation *that could have been done in advance* (by using more explicit
+forms) or other annotation (like disabling a lint). It means that minor changes
+can never leave you in a state that requires breaking changes to your own code.
```

Comment:
The "could have been done in advance" rule is a lovely formulation. It makes communicating the scope of potential breakage easy and eases the fear of wanton flaunting of semver. Well done!